image_debayer/src/image-reconstruction.cc

```diff
@@ -0,0 +1,107 @@
+  setupSubAndPub();
+}
+
+void ImageReconstruction::setupSubAndPub() {
+  // Set up the raw image subscriber.
+  boost::function<void(const sensor_msgs::ImageConstPtr&)> image_callback =
+      boost::bind(&ImageReconstruction::imageCallback, this, _1);
+  constexpr size_t kRosQueueSize = 20u;
+  sub_raw_image_ = image_transport_.subscribe(
+      FLAGS_cam_topic, kRosQueueSize, image_callback);
+
+  // Set up the processed image publisher.
+  pub_color_image_ =
+      image_transport_.advertise(FLAGS_color_topic, kRosQueueSize);
+}
+
+bool ImageReconstruction::run() {
+  LOG(INFO) << "[ImageReconstruction] Starting...";
+  spinner_.start();
+  return true;
+}
+
+void ImageReconstruction::imageCallback(
+    const sensor_msgs::ImageConstPtr& image) {
+  CHECK_NOTNULL(image);
+  cv_bridge::CvImagePtr cv_ptr;
+  try {
+    // Currently the incomplete color image output is published as a mono8.
+    if (image->encoding == sensor_msgs::image_encodings::MONO8) {
+      cv_ptr = cv_bridge::toCvCopy(image, sensor_msgs::image_encodings::MONO8);
+    } else {
+      LOG(ERROR) << "Unknown image encoding: " << image->encoding;
+    }
+  } catch (const cv_bridge::Exception& e) {  // NOLINT
+    LOG(FATAL) << "cv_bridge exception: " << e.what();
+  }
+  CHECK_NOTNULL(cv_ptr);
+  debayer(&cv_ptr->image);
+  adjustWhiteBalance(&cv_ptr->image);
+
+  publishColorImage(cv_ptr, image);
+}
+
+void ImageReconstruction::publishColorImage(
+    const cv_bridge::CvImagePtr& cv_ptr,
+    const sensor_msgs::ImageConstPtr& orig_image) {
+  CHECK_NOTNULL(cv_ptr);
+  CHECK_NOTNULL(orig_image);
+  const sensor_msgs::ImagePtr color_img_msg = cv_ptr->toImageMsg();
+  CHECK_NOTNULL(color_img_msg);
+
+  color_img_msg->encoding = sensor_msgs::image_encodings::RGB8;
+  color_img_msg->header.stamp = orig_image->header.stamp;
+  pub_color_image_.publish(color_img_msg);
+}
+
+void ImageReconstruction::debayer(cv::Mat* raw_image) {
+  CHECK_NOTNULL(raw_image);
+  cv::cvtColor(*raw_image, *raw_image, cv::COLOR_BayerGR2RGB_EA);
+}
+
+void ImageReconstruction::adjustWhiteBalance(cv::Mat* rgb_image) {
+  CHECK_NOTNULL(rgb_image);
+  // credits to http://web.stanford.edu/~sujason/ColorBalancing/simplestcb.html
+  constexpr float percent = 30;
+  constexpr float half_percent = percent / 200.0f;
+  std::vector<cv::Mat> tmpsplit(3);
+  cv::split(*rgb_image, tmpsplit);
+  for (std::size_t i = 0u; i < 3; ++i) {
+    cv::Mat flat;
+    tmpsplit[i].reshape(1, 1).copyTo(flat);
+    cv::sort(flat, flat, CV_SORT_EVERY_ROW + CV_SORT_ASCENDING);
+    const float f_cols = static_cast<float>(flat.cols);
+    const uchar lowval = flat.at<uchar>(cvFloor(f_cols * half_percent));
+    const uchar highval = flat.at<uchar>(cvCeil(f_cols * (1.0 - half_percent)));
+    tmpsplit[i].setTo(lowval, tmpsplit[i] < lowval);
+    tmpsplit[i].setTo(highval, tmpsplit[i] > highval);
+
+    cv::normalize(tmpsplit[i], tmpsplit[i], 0, 255, cv::NORM_MINMAX);
+  }
+  cv::merge(tmpsplit, *rgb_image);
```

Comment:
This is quite interesting, the results [on their site](http://web.stanford.edu/~sujason/ColorBalancing/simplestcb.html) look surprisingly good for something so simple. But it does a lot more than white balancing.
Since it truncates out the top and bottom 15% of the histogram for each color, you'd loose quite some detail in the highlights and shadows. If we intend to also use histogram equalization for brightness, choosing which correction has to be applied first might become a chicken and egg problem. E.g. adaptive equalization schemes like CLAHE might have very little effect if applied after this white balance correction, while causing weird local color shifts if applied first. If this is a problem, something like `cv::xphoto::LearningBasedWB` might help to decouple the two steps. But this is just speculation. Would be best to record a representative dataset and try :)